tortoise/tortoise.go

```diff
@@ -508,37 +525,48 @@ func (t *turtle) processLayer(ctx context.Context, logger log.Log, lid types.Lay
 	previous := t.verified
 	if t.mode.isVerifying() {
 		t.verifying.countVotes(logger, lid, ballots)
-		iterateLayers(t.verified.Add(1), t.processed.Sub(1), func(lid types.LayerID) bool {
-			ok := t.verifying.verify(logger, lid)
-			if !ok {
-				return false
-			}
-			t.verified = lid
-			updateThresholds(logger, t.Config, &t.commonState, t.mode)
-			return true
-		})
 	}
-	if t.canUseFullMode() || t.mode.isFull() {
+	for target := t.verified.Add(1); target.Before(t.processed); target = target.Add(1) {
+		var success bool
 		if t.mode.isVerifying() {
-			t.mode = t.mode.toggleMode()
-			logger.With().Info("switching to full self-healing tortoise",
-				lid,
-				log.Stringer("verified", t.verified),
-				log.Stringer("mode", t.mode),
-			)
-			// update threshold since we have different windows for vote counting in each mode
-			updateThresholds(logger, t.Config, &t.commonState, t.mode)
+			success = t.verifying.verify(logger, target)
 		}
-		t.full.countVotes(logger)
-		iterateLayers(t.verified.Add(1), t.processed.Sub(1), func(lid types.LayerID) bool {
-			ok := t.full.verify(logger, lid)
-			if !ok {
-				return false
+		if !success && (t.canUseFullMode() || t.mode.isFull()) {
+			if t.mode.isVerifying() {
+				t.switchModes(logger)
+				// reset accumulated weight as verifying tortoise will re-counting votes
+				// with input from full tortoise.
+				t.verifying.resetWeights()
 			}
-			t.verified = lid
+			t.full.countVotes(logger)
+			success = t.full.verify(logger, target)
```

Comment:
would this be easier if you make t.mode an tortoise object and so something similar to the psudocode below
```
verify:
t.mode.countVotes()
for target := t.verified.Add(1); target.Before(t.processed); target = target.Add(1) {
   success = t.mode.verify(target)
   if !success {
        if t.mode == full {
           // fail
        }
        t.swithcModes()
        goto verify  
   }
}
```

Reply:
but the pseudocode dropped almost everything, it is hard for me to tell what it will actually simplify.

restart will not be simplified, because it needs to recount votes for full distance between target + 1 and processed.
also i can't use single-mode object as i always need to have access to both.
beside that i think that goto in golang is a mistake and there is never a good reason to use it

Reply:
now that that verifying and full have the same interface (countVotes and verify), maybe it's possible to remove the frequent `if t.mode.isVerifying()` checks. the underlying mode should know what to do respectively.

just an idea. no biggie.